tests/test_worker.py

```diff
@@ -649,6 +654,34 @@ def test_working_worker_cold_shutdown(self):
         self.assertIsNotNone(shutdown_requested_date)
         self.assertEqual(type(shutdown_requested_date).__name__, 'datetime')
 
+    @slow
+    def test_work_horse_death_sets_job_failed(self):
+        """worker with an ongoing job whose work horse dies unexpectadly should
+        set the job's status either to FINISHED or FAILED
+        """
+        fooq = Queue('foo')
+        failed_q = get_failed_queue()
+        self.assertEqual(failed_q.count, 0)
+        self.assertEqual(fooq.count, 0)
+        w = Worker(fooq)
+        sentinel_file = '/tmp/.rq_sentinel_work_horse_death'
+        if os.path.exists(sentinel_file):
+            os.remove(sentinel_file)
+        fooq.enqueue(create_file_after_timeout, sentinel_file, 100)
+        job, queue = w.dequeue_job_and_maintain_ttl(5)
+        w.fork_work_horse(job, queue)
+        p = Process(target=wait_and_kill_work_horse, args=(w._horse_pid, 0.5))
+        p.start()
+        w.monitor_work_horse(job)
+        job_status = job.get_status()
+        p.join(1)
+        if os.path.exists(sentinel_file):
```

Comment:
My reasoning was that in the ***extreme*** case where the **create_file_after_timeout** job might run to completion before **wait_and_kill_work_horse** gets to kill it (think extremely high system load or something killing the **wait_and_kill_work_horse** process), in which case we don't get to actually test what we are interested in, the test case should succeed (we did not find a failure) than randomly fail. I was probably being overly defensive, since in general I don't expect it to ever hit the if part of that if-else. If you want I can remove it.